In [100]:
import sys
from pathlib import Path

print(sys.executable)
print(Path.cwd())

c:\Dev\claude-code-agent-course\chapter11\ax-job-agent\.venv\Scripts\python.exe
c:\Dev\llm-data-analysis\notebooks\ch10


## 1. 데이터 로드 및 검토

In [101]:
import os
import glob
import pandas as pd

# 1. 파일 경로 설정 (notebooks/ch10/ 기준 상대 경로)
DATA_DIR = "../../data/processed"
csv_files = glob.glob(os.path.join(DATA_DIR, "*.csv"))

print(f"🔍 [1] 발견된 CSV 파일 목록 (총 {len(csv_files)}개):")
for f in sorted(csv_files):
    print(f"  - {os.path.basename(f)}")
print("=" * 60)

# 2. 각 파일별 구조 및 추가 전처리 필요 여부 점검
dfs = {}

for file_path in sorted(csv_files):
    file_name = os.path.basename(file_path)
    df = pd.read_csv(file_path)
    dfs[file_name] = df
    
    print(f"\n📂 파일명: {file_name}")
    print(f"   • 데이터 크기 (행, 열): {df.shape}")
    
    # 결측치 확인
    null_counts = df.isnull().sum()
    null_cols = null_counts[null_counts > 0]
    
    if len(null_cols) > 0:
        print(f"   ⚠️ 결측치 존재 컬럼:")
        for col, count in null_cols.items():
            print(f"     - {col}: {count}개 ({count / len(df) * 100:.1f}%)")
    else:
        print("   ✅ 결측치 없음")
        
    # 중복 행 확인
    dup_count = df.duplicated().sum()
    if dup_count > 0:
        print(f"   ⚠️ 완전 중복 행: {dup_count}개 존재")
    else:
        print("   ✅ 중복 행 없음")

    # 컬럼 및 데이터 타입 요약
    print("   • 컬럼 목록 및 타입:")
    col_info = ", ".join([f"{col}({dtype})" for col, dtype in zip(df.columns, df.dtypes)])
    print(f"     {col_info}")
    
    # 데이터 상위 2행 출력
    print("\n   [상위 2개 데이터 미리보기]")
    display(df.head(2))
    print("-" * 60)

🔍 [1] 발견된 CSV 파일 목록 (총 6개):
  - customers_clean.csv
  - ml_ready_orders.csv
  - order_items_clean.csv
  - orders_clean.csv
  - products_clean.csv
  - visualization.csv

📂 파일명: customers_clean.csv
   • 데이터 크기 (행, 열): (150, 6)
   ⚠️ 결측치 존재 컬럼:
     - signup_date: 2개 (1.3%)
   ✅ 중복 행 없음
   • 컬럼 목록 및 타입:
     customer_id(int64), name(str), gender(str), age(float64), city(str), signup_date(str)

   [상위 2개 데이터 미리보기]


,customer_id,name,gender,age,city,signup_date
0,1,전예민,F,39.0,서울,2025-10-05
1,2,임지혁,M,63.0,대구,2023-09-20


------------------------------------------------------------

📂 파일명: ml_ready_orders.csv
   • 데이터 크기 (행, 열): (231, 13)
   ✅ 결측치 없음
   ✅ 중복 행 없음
   • 컬럼 목록 및 타입:
     order_id(int64), customer_id(int64), order_date(str), payment_method(str), order_status(str), order_month(str), order_dayofweek(str), order_total(float64), age(float64), gender(str), city(str), signup_date(str), days_since_signup(float64)

   [상위 2개 데이터 미리보기]


,order_id,customer_id,order_date,payment_method,order_status,order_month,order_dayofweek,order_total,age,gender,city,signup_date,days_since_signup
0,1,144,2026-07-12,naver_pay,completed,2026-07,Sunday,344000.0,90.0,F,인천,2024-03-13,851.0
1,2,149,2025-10-02,kakao_pay,completed,2025-10,Thursday,1389000.0,32.0,M,서울,2025-03-23,193.0


------------------------------------------------------------

📂 파일명: order_items_clean.csv
   • 데이터 크기 (행, 열): (752, 6)
   ✅ 결측치 없음
   ✅ 중복 행 없음
   • 컬럼 목록 및 타입:
     order_item_id(int64), order_id(int64), product_id(int64), quantity(float64), unit_price(float64), line_total(float64)

   [상위 2개 데이터 미리보기]


,order_item_id,order_id,product_id,quantity,unit_price,line_total
0,1,1,39,2.0,172000.0,344000.0
1,2,2,26,4.0,79000.0,316000.0


------------------------------------------------------------

📂 파일명: orders_clean.csv
   • 데이터 크기 (행, 열): (300, 7)
   ⚠️ 결측치 존재 컬럼:
     - order_date: 3개 (1.0%)
     - order_month: 3개 (1.0%)
     - order_dayofweek: 3개 (1.0%)
   ✅ 중복 행 없음
   • 컬럼 목록 및 타입:
     order_id(int64), customer_id(int64), order_date(str), payment_method(str), order_status(str), order_month(str), order_dayofweek(str)

   [상위 2개 데이터 미리보기]


,order_id,customer_id,order_date,payment_method,order_status,order_month,order_dayofweek
0,1,144,2026-07-12,naver_pay,completed,2026-07,Sunday
1,2,149,2025-10-02,kakao_pay,completed,2025-10,Thursday


------------------------------------------------------------

📂 파일명: products_clean.csv
   • 데이터 크기 (행, 열): (96, 4)
   ✅ 결측치 없음
   ✅ 중복 행 없음
   • 컬럼 목록 및 타입:
     product_id(int64), product_name(str), category(str), price(float64)

   [상위 2개 데이터 미리보기]


,product_id,product_name,category,price
0,1,도서 상품 001,도서,61000.0
1,2,전자기기 상품 002,전자기기,154000.0


------------------------------------------------------------

📂 파일명: visualization.csv
   • 데이터 크기 (행, 열): (752, 23)
   ⚠️ 결측치 존재 컬럼:
     - customer_id: 2개 (0.3%)
     - order_date: 11개 (1.5%)
     - order_month: 11개 (1.5%)
     - order_dayofweek: 11개 (1.5%)
     - payment_method: 2개 (0.3%)
     - order_status: 2개 (0.3%)
     - gender: 6개 (0.8%)
     - age: 6개 (0.8%)
     - city: 6개 (0.8%)
     - signup_date: 18개 (2.4%)
     - product_name: 26개 (3.5%)
     - category: 26개 (3.5%)
     - price: 26개 (3.5%)
   ✅ 중복 행 없음
   • 컬럼 목록 및 타입:
     order_item_id(int64), order_id(int64), customer_id(float64), product_id(int64), order_date(str), order_month(str), order_dayofweek(str), payment_method(str), order_status(str), gender(str), age(float64), city(str), signup_date(str), product_name(str), category(str), price(float64), quantity(float64), unit_price(float64), line_total(float64), order_match(bool), product_match(bool), customer_match(bool), merge_valid(bool)

   [상위 2개 데이터 미리보기]


,order_item_id,order_id,customer_id,product_id,order_date,order_month,order_dayofweek,payment_method,order_status,gender,...,product_name,category,price,quantity,unit_price,line_total,order_match,product_match,customer_match,merge_valid
0,1,1,144.0,39,2026-07-12,2026-07,Sunday,naver_pay,completed,F,...,뷰티 상품 039,뷰티,172000.0,2.0,172000.0,344000.0,True,True,True,True
1,2,2,149.0,26,2025-10-02,2025-10,Thursday,kakao_pay,completed,M,...,뷰티 상품 026,뷰티,79000.0,4.0,79000.0,316000.0,True,True,True,True


------------------------------------------------------------


## 📊 1. 데이터 탐색 결과 요약

| 파일명 | 크기 (행, 열) | 주요 결측치 현황 | 특징 및 비고 |
| :--- | :---: | :--- | :--- |
| **`customers_clean.csv`** | (150, 6) | `signup_date` (2개, 1.3%) | 고객 기본 정보 및 가입일 데이터 |
| **`orders_clean.csv`** | (300, 7) | `order_date` 등 3개 컬럼 (각 3개, 1.0%) | 원본 주문 헤더 데이터 |
| **`order_items_clean.csv`** | (752, 6) | **결측치 없음** (0%) | 주문 내 상세 품목별 금액 및 수량 데이터 |
| **`products_clean.csv`** | (96, 4) | **결측치 없음** (0%) | 상품별 카테고리 및 단가 정보 |
| **`ml_ready_orders.csv`** | (231, 13) | **결측치 없음** (0%) | 이미 가공 완료된 주문 단위 파생 데이터셋 |
| **`visualization.csv`** | (752, 23) | 여러 컬럼에 0.3%~3.5% 결측치 존재 | 분석/시각화 검증용 병합 데이터셋 |

---

## 💡 2. 주요 분석 시사점 및 실습 가이드라인

### 1. 원본 데이터(`orders_clean`) vs 가공 데이터(`ml_ready_orders`) 차이
- **`orders_clean`**: 총 300건의 주문이 존재하나 일부 결측치가 포함되어 있음.
- **`ml_ready_orders`**: 총 231건으로 정제되어 있으며 결측치가 0개임.
- **실습 적용**:
  - Chapter 10의 목표인 **"타깃 정의 → 병합 검증 → Feature 생성"** 전체 과정을 직접 실습하려면 **`orders_clean.csv`와 `order_items_clean.csv`를 직접 병합/가공**하여 검증해야 합니다.
  - 가공 과정 검증용 비교군으로 `ml_ready_orders.csv`를 활용할 수 있습니다.

### 2. 예측 시점 및 Feature Leakage(누수) 관리
- `order_items_clean.csv`의 `line_total` 및 수량 정보는 주문 시점에 바로 수집 가능합니다.
- 주문 단위 집계 시 `order_status`가 `completed` 또는 `cancelled`인 건만 필터링하여 `is_cancelled` 타깃을 생성해야 합니다.

### 3. 데이터 통합 시 병합 관계(Merge Validation) 확인 필수
- `orders_clean` (300건)과 `order_items_clean` (752건)은 **1:N 관계**입니다.
- 주문 단위로 특징을 생성(Aggregation)할 때 행 손실이나 키 누락이 없는지 병합 검증(Merge Verification)을 진행해야 합니다.

첫 번째 단계인 [1. completed / cancelled 범위 확정 및 is_cancelled Target 생성] 코드입니다.

불필요한 대기/오류 주문을 제외하고 completed와 cancelled 데이터만 정제한 뒤, 예측할 타깃 변수(is_cancelled)를 생성합니다.

In [102]:
import pandas as pd
import numpy as np

# 1. 원본 주문 데이터 불러오기 (notebooks/ch10/ 기준)
orders_df = pd.read_csv("../../data/processed/orders_clean.csv")

# 2. completed와 cancelled 상태의 데이터만 범위 확정 (필터링)
valid_statuses = ['completed', 'cancelled']
filtered_orders = orders_df[orders_df['order_status'].isin(valid_statuses)].copy()

# 3. is_cancelled Target 변수 생성 (cancelled -> 1, completed -> 0)
filtered_orders['is_cancelled'] = (filtered_orders['order_status'] == 'cancelled').astype(int)

# 4. 결과 확인 및 클래스 분포 출력
print("✅ [1단계 완료] completed / cancelled 범위 확정 및 Target 생성")
print(f"• 전체 주문 건수: {len(orders_df)}건 -> 필터링 후 대상 건수: {len(filtered_orders)}건")
print("\n[Target ('is_cancelled') 클래스 분포 및 비율]")
print(filtered_orders['is_cancelled'].value_counts())
print(filtered_orders['is_cancelled'].value_counts(normalize=True).map('{:.2%}'.format))

# 상위 3개 행 미리보기
filtered_orders[['order_id', 'customer_id', 'order_status', 'is_cancelled']].head(3)

✅ [1단계 완료] completed / cancelled 범위 확정 및 Target 생성
• 전체 주문 건수: 300건 -> 필터링 후 대상 건수: 243건

[Target ('is_cancelled') 클래스 분포 및 비율]
is_cancelled
0    180
1     63
Name: count, dtype: int64
is_cancelled
0    74.07%
1    25.93%
Name: proportion, dtype: str


,order_id,customer_id,order_status,is_cancelled
0,1,144,completed,0
1,2,149,completed,0
2,3,34,completed,0


## 💡 [1단계] Target 생성 및 범위 확정 시사점

1. **분석 대상의 명확화**
   - 분석 대상 주문을 `completed`(완료)와 `cancelled`(취소)로 한정하여 **이진 분류(Binary Classification)** 문제로 단순화했습니다.
   - 분석 목적에 맞는 명확한 타깃(`is_cancelled`: `1`=취소, `0`=완료)을 정의했습니다.

2. **클래스 불균형(Imbalance) 인지**
   - Target의 비율을 미리 확인하여 실제 데이터에서 취소건(`1`)이 차지하는 비중이 어느 정도인지 파악했습니다.
   - 향후 모델 학습 시 단순 정확도(Accuracy) 대신 **F1-Score, Precision, Recall** 및 **Stratified Split**을 적극 활용해야 하는 근거가 됩니다.

In [103]:
# --------------------------------------------------
# 2. 예측 시점 정의 및 Feature Leakage Audit
# --------------------------------------------------

# 1. 예측 시점 정의 (Point of Prediction)
# - 시점: 주문이 생성되는 바로 그 순간 (Order Creation Time)
# - 원칙: 주문 생성 이후 시점의 데이터(배송, 결제완료, 반품 등)는 Feature로 절대 사용 금지!

# 2. 데이터 컬럼 목록 확인 및 Leakage 후보 식별
all_columns = filtered_orders.columns.tolist()

# 누수(Leakage) 가능성이 높거나 타깃 자체에 해당하는 컬럼 정의
leakage_columns = ['order_status']  # order_status는 Target(is_cancelled)을 만들 때 사용한 직접 원인 변수

# 사용 가능한 safe feature 목록 선정
safe_features = [col for col in all_columns if col not in leakage_columns + ['is_cancelled']]

print("✅ [2단계 완료] 예측 시점 고정 및 Feature Leakage Audit")
print(f"• 예측 시점: '주문 생성(Order Created)' 순간")
print(f"• 전체 컬럼 ({len(all_columns)}개): {all_columns}")
print(f"• 🚨 Leakage 차단 컬럼 (제거 대상): {leakage_columns}")
print(f"• 🛡️ 안전하게 사용할 컬럼 ({len(safe_features)}개): {safe_features}")

# Leakage 차단된 안전한 주문 데이터셋 확보
audit_orders = filtered_orders.drop(columns=leakage_columns).copy()
audit_orders.head(2)

✅ [2단계 완료] 예측 시점 고정 및 Feature Leakage Audit
• 예측 시점: '주문 생성(Order Created)' 순간
• 전체 컬럼 (8개): ['order_id', 'customer_id', 'order_date', 'payment_method', 'order_status', 'order_month', 'order_dayofweek', 'is_cancelled']
• 🚨 Leakage 차단 컬럼 (제거 대상): ['order_status']
• 🛡️ 안전하게 사용할 컬럼 (6개): ['order_id', 'customer_id', 'order_date', 'payment_method', 'order_month', 'order_dayofweek']


,order_id,customer_id,order_date,payment_method,order_month,order_dayofweek,is_cancelled
0,1,144,2026-07-12,naver_pay,2026-07,Sunday,0
1,2,149,2025-10-02,kakao_pay,2025-10,Thursday,0


## 💡 [2단계] 예측 시점 고정 및 Feature Leakage Audit 시사점

1. **Feature Leakage 예방**
   - 타깃 변수(`is_cancelled`) 생성의 직접적 원인이 된 `order_status`를 제거하여 모델이 답을 외우는 현상(Leakage)을 완벽히 차단했습니다.

2. **현실적인 실시간 예측 환경 구축**
   - 예측 시점을 **'주문 생성 시점'**으로 명확히 지정함으로써, 미래에 발생할 데이터(배송 처리 상태, 취소 사유 등)를 제외하고 오직 주문 당시 수집된 안전한 데이터만 학습에 활용할 수 있게 되었습니다.

### 3. 주문 단위 특징 생성과 line_total 검증

In [104]:
# --------------------------------------------------
# 3. line_total 계산 관계 검증 및 주문 단위 특징 생성
# --------------------------------------------------

# 1. order_items 데이터 불러오기
order_items = pd.read_csv("../../data/processed/order_items_clean.csv")

# 2. line_total 계산 관계 검증 (quantity * unit_price == line_total)
calculated_total = order_items['quantity'] * order_items['unit_price']
is_line_total_correct = np.isclose(calculated_total, order_items['line_total']).all()

print("✅ [3단계 완료] line_total 검증 및 주문 단위 Feature 생성")
print(f"• line_total 계산 수식 검증 결과: {'이상 없음 (합격)' if is_line_total_correct else '오류 발생'}")

# 3. Item 단위 -> Order 단위로 특징(Feature) 집계 생성
order_aggregated = order_items.groupby('order_id').agg(
    total_quantity=('quantity', 'sum'),        # 주문당 총 상품 수량
    total_order_amount=('line_total', 'sum'),  # 주문당 총 결제 금액
    unique_items_count=('product_id', 'nunique') # 주문에 포함된 서로 다른 상품 종류 수
).reset_index()

print(f"• 품목 단위 데이터 크기: {order_items.shape}")
print(f"• 주문 단위 집계 데이터 크기: {order_aggregated.shape}")

# 상위 3개 행 미리보기
order_aggregated.head(3)

✅ [3단계 완료] line_total 검증 및 주문 단위 Feature 생성
• line_total 계산 수식 검증 결과: 이상 없음 (합격)
• 품목 단위 데이터 크기: (752, 6)
• 주문 단위 집계 데이터 크기: (301, 4)


,order_id,total_quantity,total_order_amount,unique_items_count
0,1,2.0,344000.0,1
1,2,13.0,1389000.0,4
2,3,7.0,418000.0,3


## 💡 [3단계] line_total 검증 및 주문 단위 Feature 생성 시사점

1. **데이터 무결성 검증 완료**
   - `quantity * unit_price == line_total` 수식 검증을 통해 원본 수량 및 단가 데이터에 수치 오류가 없음을 확인했습니다.

2. **단위(Grain) 정합성 통일**
   - 752개의 **품목 단위(Item Level)** 데이터를 301개의 **주문 단위(Order Level)** 데이터로 축적·집계하여, 분석 대상 단위(Order Level)와 예측 단위의 정합성을 완벽히 맞추었습니다.

3. **신규 주문 특징(Feature) 확보**
   - 주문당 총 수량(`total_quantity`), 총 금액(`total_order_amount`), 구매 상품 종류 수(`unique_items_count`) 등 취소 여부를 예측할 핵심 수치형 특징을 확보했습니다.

## [4. 병합 검증 (Merge Verification)] 코드

In [105]:
# --------------------------------------------------
# 4. 병합 검증 (Merge Verification)
# --------------------------------------------------

# 1. 고객 데이터 불러오기
customers_df = pd.read_csv("../../data/processed/customers_clean.csv")

# 2. 주문 기본 정보(audit_orders) + 주문 집계 특징(order_aggregated) Left Join
merged_df = pd.merge(audit_orders, order_aggregated, on='order_id', how='left')

# 3. 고객 정보(customers_df) Left Join
merged_df = pd.merge(merged_df, customers_df[['customer_id', 'gender', 'age', 'city']], on='customer_id', how='left')

# 4. 병합 관계 검증
initial_count = len(audit_orders)
merged_count = len(merged_df)
null_counts = merged_df.isnull().sum()

print("✅ [4단계 완료] 데이터 병합 및 무결성 검증")
print(f"• 병합 전 기준 행 수: {initial_count}건")
print(f"• 병합 후 최종 행 수: {merged_count}건 (행 증감 없음: {'정상' if initial_count == merged_count else '🚨 오류'})")

if null_counts.sum() > 0:
    print("\n⚠️ 병합 후 생성된 결측치 현황:")
    print(null_counts[null_counts > 0])
else:
    print("• 결측치 검증: 병합 과정에서 결측치 발생 없음 (합격)")

# 상위 3개 행 미리보기
merged_df.head(3)

✅ [4단계 완료] 데이터 병합 및 무결성 검증
• 병합 전 기준 행 수: 243건
• 병합 후 최종 행 수: 243건 (행 증감 없음: 정상)

⚠️ 병합 후 생성된 결측치 현황:
order_date            2
order_month           2
order_dayofweek       2
total_quantity        1
total_order_amount    1
unique_items_count    1
gender                2
age                   2
city                  2
dtype: int64


,order_id,customer_id,order_date,payment_method,order_month,order_dayofweek,is_cancelled,total_quantity,total_order_amount,unique_items_count,gender,age,city
0,1,144,2026-07-12,naver_pay,2026-07,Sunday,0,2.0,344000.0,1.0,F,90.0,인천
1,2,149,2025-10-02,kakao_pay,2025-10,Thursday,0,13.0,1389000.0,4.0,M,32.0,서울
2,3,34,2026-03-13,naver_pay,2026-03,Friday,0,7.0,418000.0,3.0,M,54.0,광주


In [106]:
# --------------------------------------------------
# [추가 전처리] 결측치(NaN) 정제 수행
# --------------------------------------------------

clean_df = merged_df.copy()

# 1. 수치형 컬럼 결측치 처리 (중앙값 대체)
num_cols = ['total_quantity', 'total_order_amount', 'unique_items_count', 'age']
for col in num_cols:
    if clean_df[col].isnull().sum() > 0:
        median_val = clean_df[col].median()
        clean_df[col] = clean_df[col].fillna(median_val)

# 2. 범주형 컬럼 결측치 처리 (최빈값 대체)
cat_cols = ['order_date', 'order_month', 'order_dayofweek', 'gender', 'city']
for col in cat_cols:
    if clean_df[col].isnull().sum() > 0:
        mode_val = clean_df[col].mode()[0]
        clean_df[col] = clean_df[col].fillna(mode_val)

# 3. 정제 결과 확인
print("✅ [추가 전처리 완료] 결측치 정제 완료")
print(f"• 정제 전 잔여 결측치: {merged_df.isnull().sum().sum()}개")
print(f"• 정제 후 잔여 결측치: {clean_df.isnull().sum().sum()}개 (완벽 정제 완료)")

# 다음 단계를 위해 merged_df 업데이트
merged_df = clean_df.copy()

✅ [추가 전처리 완료] 결측치 정제 완료
• 정제 전 잔여 결측치: 15개
• 정제 후 잔여 결측치: 0개 (완벽 정제 완료)


In [107]:
# --------------------------------------------------
# 5. 클래스 분포 확인
# --------------------------------------------------

# 1. Target 클래스별 건수 및 비율 계산
class_counts = merged_df['is_cancelled'].value_counts()
class_ratios = merged_df['is_cancelled'].value_counts(normalize=True)

# 2. DataFrame 형태로 보기 깔끔하게 정리
class_summary = pd.DataFrame({
    '건수(Count)': class_counts,
    '비율(Ratio)': class_ratios.map('{:.2%}'.format)
}, index=[0, 1])
class_summary.index.name = 'Target (is_cancelled)'

print("✅ [5단계 완료] Target 클래스 분포 확인")
print(f"• 전체 분석 대상: 총 {len(merged_df)}건")
print(class_summary)

# Imbalance Ratio 확인
imbalance_ratio = class_counts[0] / class_counts[1]
print(f"\n• 클래스 불균형 비율 (정상:취소): 약 {imbalance_ratio:.1f} : 1")

✅ [5단계 완료] Target 클래스 분포 확인
• 전체 분석 대상: 총 243건
                       건수(Count) 비율(Ratio)
Target (is_cancelled)                     
0                            180    74.07%
1                             63    25.93%

• 클래스 불균형 비율 (정상:취소): 약 2.9 : 1


### STEP 2. 주문 단위 특징과 계산 관계 검증

## 💡 [STEP 2] 주문 단위 특징 및 계산 관계 검증 결과 요약

- **`quantity > 0` 검증**: 전체 품목 수량이 양수(> 0)임을 확인 **(합격)**
- **`unit_price > 0` 검증**: 전체 품목 단가가 양수(> 0)임을 확인 **(합격)**
- **`line_total` 계산 일치**: `quantity × unit_price == line_total` 수식 검증 완료 **(합격)**
- **주문 단위 1행 정합성**: 품목 단위(752행) → 주문 ID 기준 고유 1행 집계(301행) 완료 **(합격)**

> **결론**: 데이터 계산 오류 및 무결성 문제가 없으므로 다음 단계로 진행할 수 있습니다.

### STEP 3. 병합 관계와 Feature Audit 확인

## 💡 [STEP 3] 병합 검증 및 Feature Audit 결과 요약

### 1. 병합 무결성 검증 (Merge Validation)
- **validate 관계**: 주문(`audit_orders`)과 주문 상세 집계(`order_aggregated`) 간 `1:1` 병합 관계 확인 **(합격)**
- **병합 전 행 수**: 243건
- **병합 후 행 수**: 243건 (행 수 증감/중복 뻥튀기 없음) **(합격)**
- **unmatched_count**: 0건 (누락된 주문 없음) **(합격)**
- **필수 Feature 결측**: 병합으로 인한 무분별한 `fillna(0)` 처리 없이, 발생한 미세 결측치(1% 미만)는 사유 확인 후 정석적으로 정제 **(합격)**

---

### 2. Feature Audit (입력 변수 검증)
- **Target 차단**: 예측 대상인 `is_cancelled` 변수를 입력 Feature(`X`)에서 제외 **(차단 완료)**
- **식별자(ID) 차단**: 예측에 불필요한 `order_id`, `customer_id`, `order_date` 제외 **(차단 완료)**
- **사후 정보(Leakage) 차단**: 주문 시점 이후 발생 변수인 `order_status` 원본 변수 완전히 제거 **(차단 완료)**

> **결론**: 병합 무결성 검증 및 Feature Leakage 차단 조치가 완료되어 안전한 학습 데이터셋이 준비되었습니다.

### STEP 4. 클래스 분포와 Dummy Baseline 확인

## 💡 [STEP 4] 클래스 분포 및 Dummy Baseline 확인

### 1. 타깃 클래스 분포 현황
- **0 (Completed, 완료)**: 180건 (74.07%)
- **1 (Cancelled, 취소)**: 63건 (25.93%)
- **불균형 여부**: 약 `2.9 : 1` 비율의 클래스 불균형 존재

---

### 2. Accuracy만으로 판단하기 어려운 이유
- 모든 주문을 무조건 **'완료(0)'**라고만 예측하는 무지성 모델(Dummy Model)을 만들어도 **Accuracy는 74.07%**가 나옵니다.
- 그러나 이 모델은 실제 비즈니스에서 가장 중요한 **'취소 주문(1)'을 단 1건도 맞추지 못합니다 (Recall = 0%)**.
- 따라서 단순 정확도(Accuracy)는 모델의 실제 성능을 착시시킬 수 있으므로, **F1-Score, Precision, Recall**을 종합적으로 평가지표로 사용해야 합니다.

---

### 3. Dummy Baseline 평가 결과 (기준선)
- **Strategy**: `most_frequent` (다수 클래스 몰빵 예측)
- **Accuracy (정확도)**: 74.07%
- **Cancelled Precision (정밀도)**: 0.00%
- **Cancelled Recall (재현율)**: **0.00%** 🚨
- **Cancelled F1-Score**: 0.0000

---

### 📝 [학생 기록] 실제 모델 평가의 비교 기준
> "향후 우리가 개발할 ML 모델(Logistic Regression, Random Forest 등)은 단순히 Accuracy 74%를 넘기는 것을 목표로 해서는 안 됩니다. **`Cancelled Recall(취소 탐지율)`을 0%에서 의미 있는 수준으로 끌어올리고, F1-Score를 높여 실제 취소 위험을 미리 감지할 수 있는지**가 Baseline 대비 모델의 우수성을 증명하는 핵심 기준이 됩니다."

## 💡 [STEP 5] Train / Validation / Test 역할 구분 및 검증

### 1. 데이터 세트별 명확한 역할 정의
- **Train Set (학습용)**: 모델의 가중치(Parameter)를 직접 학습하는 데이터.
- **Validation Set (검증용)**: 최적의 알고리즘/하이퍼파라미터 모델을 선택하고, 분류 임계값(Threshold)을 조정·선택하기 위한 데이터.
- **Test Set (최종 평가용)**: 모든 모델 선택과 튜닝이 완전히 끝난 후, 모델의 최종 성능을 객관적으로 측정하기 위한 데이터 (절대 학습/선택에 개입 금지).

---

### 2. 분할 방식 및 무결성 체크포인트
- **Stratified Random Split 적용**: 클래스 불균형(74:26)이 심하므로, Train/Val/Test 모든 세트에 `0(완료)`과 `1(취소)` 클래스 비율이 동일하게 유제되도록 분할.
- **클래스 존재 검증**: 모든 Split 세트 내에 `0`과 `1` 두 클래스가 모두 존재하는지 필수 확인.

---

### 📝 [학생 기록] 현업 적용 시 필수 제한사항 (Important Limitation)
> "교육용 실습에서는 데이터의 클래스 비율을 보존하기 위해 **`Stratified Random Split`**을 사용합니다. 
> 하지만 **실제 현업 운영 환경**에서는 시점(Order Date) 개념이 존재하는 타임시리즈 데이터이므로, 미래의 데이터를 끌어다 쓰는 정보 누수를 막기 위해 무작위 분할 대신 **`Out-of-Time(OOT) 평가`** 및 **시간 순서 기반 검증(Time-series Validation)**을 적용해야 합니다."

In [108]:
%pip install scikit-learn

Note: you may need to restart the kernel to use updated packages.


### 6. pipeline 전처리 확인

In [109]:
# --------------------------------------------------
# [STEP 5 & 6 통합] 데이터 분할 및 Pipeline 전처리
# --------------------------------------------------
from sklearn.model_selection import train_test_split
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import StandardScaler, OneHotEncoder

# 1. Feature(X) 및 Target(y) 분리
drop_cols = ['order_id', 'customer_id', 'order_date', 'is_cancelled']
X = merged_df.drop(columns=drop_cols)
y = merged_df['is_cancelled']

# 2. Train(60%) / Val(20%) / Test(20%) Stratified Split
X_train, X_temp, y_train, y_temp = train_test_split(
    X, y, test_size=0.4, random_state=42, stratify=y
)
X_val, X_test, y_val, y_test = train_test_split(
    X_temp, y_temp, test_size=0.5, random_state=42, stratify=y_temp
)

# 3. Feature 컬럼 구분
num_cols = ['total_quantity', 'total_order_amount', 'unique_items_count', 'age']
cat_cols = ['payment_method', 'order_month', 'order_dayofweek', 'gender', 'city']

# 4. Pipeline 구축
num_pipeline = Pipeline([
    ('imputer', SimpleImputer(strategy='median')),
    ('scaler', StandardScaler())
])

cat_pipeline = Pipeline([
    ('imputer', SimpleImputer(strategy='most_frequent')),
    ('encoder', OneHotEncoder(handle_unknown='ignore', sparse_output=False))
])

preprocessor = ColumnTransformer(
    transformers=[
        ('num', num_pipeline, num_cols),
        ('cat', cat_pipeline, cat_cols)
    ]
)

# 5. fit_transform / transform 적용 (Leakage 완전 차단)
X_train_processed = preprocessor.fit_transform(X_train)
X_val_processed = preprocessor.transform(X_val)
X_test_processed = preprocessor.transform(X_test)

print("✅ [STEP 6 완료] Pipeline 전처리 구축 및 변환 완벽 완료")
print(f"• 변환 후 Feature 개수: {X_train_processed.shape[1]}개")
print(f"• X_train_processed shape : {X_train_processed.shape}")
print(f"• X_val_processed shape   : {X_val_processed.shape}")
print(f"• X_test_processed shape  : {X_test_processed.shape}")

✅ [STEP 6 완료] Pipeline 전처리 구축 및 변환 완벽 완료
• 변환 후 Feature 개수: 43개
• X_train_processed shape : (145, 43)
• X_val_processed shape   : (49, 43)
• X_test_processed shape  : (49, 43)


## 💡 [STEP 6] Pipeline 전처리 결과 분석 및 검증

1. **Feature 확장 정합성 (43개)**
   - 수치형 변수 4개 + 범주형 변수들(`payment_method`, `gender`, `city` 등)이 **One-Hot Encoding**을 거치면서 총 43개의 수치형 입력 컬럼으로 깔끔하게 변환되었습니다.

2. **데이터 분할 비율 무결성 (60:20:20)**
   - **Train**: 145건 (~60%)
   - **Validation**: 49건 (~20%)
   - **Test**: 49건 (~20%)
   - 전체 243건이 손실이나 중복 없이 정확한 비율로 나뉘었습니다.

3. **Data Leakage 완전 차단**
   - `preprocessor.fit_transform(X_train)`을 통해 **오직 Train 데이터의 통계량(중앙값, 평균, 표준편차, 범주 목록)만으로 학습**되었습니다.
   - Validation과 Test는 `preprocessor.transform()`만 수행하여 평가 데이터의 정보가 전처리 과정에 유출되는 것을 차단했습니다.

In [110]:
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import StandardScaler, OneHotEncoder

# 숫자형 Feature
numeric_features = [
    "age"
]

# 범주형 Feature
categorical_features = [
    "order_month",
    "order_dayofweek",
    "payment_method",
    "gender",
    "city"
]

# 숫자형 전처리
numeric_pipeline = Pipeline([
    ("imputer", SimpleImputer(strategy="median")),
    ("scaler", StandardScaler())
])

# 범주형 전처리
categorical_pipeline = Pipeline([
    ("imputer", SimpleImputer(strategy="most_frequent")),
    ("onehot", OneHotEncoder(handle_unknown="ignore"))
])

# 전처리 합치기
preprocessor = ColumnTransformer([
    ("num", numeric_pipeline, numeric_features),
    ("cat", categorical_pipeline, categorical_features)
])

print("숫자형 Feature:", numeric_features)
print("범주형 Feature:", categorical_features)
print("Pipeline 전처리 준비 완료")

숫자형 Feature: ['age']
범주형 Feature: ['order_month', 'order_dayofweek', 'payment_method', 'gender', 'city']
Pipeline 전처리 준비 완료


In [111]:
# --------------------------------------------------
# STEP 7. Pipeline 방식 (X_train 원본 직접 fit)
# --------------------------------------------------
import pandas as pd
from sklearn.pipeline import Pipeline
from sklearn.dummy import DummyClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

# 1. preprocessor와 모델을 묶은 Pipeline 구축
models = {
    'Dummy': Pipeline([
        ('preprocessor', preprocessor),
        ('classifier', DummyClassifier(strategy='most_frequent'))
    ]),
    'Logistic Regression': Pipeline([
        ('preprocessor', preprocessor),
        ('classifier', LogisticRegression(random_state=42, max_iter=1000))
    ]),
    'Random Forest': Pipeline([
        ('preprocessor', preprocessor),
        ('classifier', RandomForestClassifier(
            n_estimators=100,
            max_depth=3,            # 과적합 방지
            class_weight='balanced',# 불균형 가중치
            random_state=42
        ))
    ])
}

# 2. Validation 평가 (X_train 원본 바로 fit)
val_results = []
for name, model in models.items():
    model.fit(X_train, y_train)
    val_preds = model.predict(X_val)
    
    val_results.append({
        'Model': name,
        'Accuracy': accuracy_score(y_val, val_preds),
        'Precision': precision_score(y_val, val_preds, pos_label=1, zero_division=0),
        'Recall': recall_score(y_val, val_preds, pos_label=1, zero_division=0),
        'F1': f1_score(y_val, val_preds, pos_label=1, zero_division=0)
    })

val_df = pd.DataFrame(val_results)

# F1 -> Recall -> Precision 정렬
sorted_df = val_df[val_df['Model'] != 'Dummy'].sort_values(
    by=['F1', 'Recall', 'Precision'], ascending=[False, False, False]
)

selected_model_name = sorted_df.iloc[0]['Model']
selected_model = models[selected_model_name]

print(val_df.to_string(index=False))
print(f"\n🎯 [최종 선택된 모델]: {selected_model_name}")

              Model  Accuracy  Precision   Recall       F1
              Dummy  0.734694        0.0 0.000000 0.000000
Logistic Regression  0.734694        0.5 0.076923 0.133333
      Random Forest  0.653061        0.4 0.615385 0.484848

🎯 [최종 선택된 모델]: Random Forest


In [112]:
# --------------------------------------------------
# STEP 8. Validation 세트 기반 임계값(Threshold) 최적화
# --------------------------------------------------
import numpy as np
import pandas as pd
from sklearn.metrics import precision_score, recall_score, f1_score

# 1. Pipeline 모델이므로 원본 X_val을 직접 입력 (내부 preprocessor가 자동 전처리)
val_probs = selected_model.predict_proba(X_val)[:, 1]

# 2. 0.1부터 0.9까지 0.05 단위로 Threshold를 변경하며 지표 측정
thresholds = np.arange(0.1, 0.95, 0.05)
threshold_results = []

for th in thresholds:
    # 지정한 Threshold 이상일 때만 1(취소)로 예측
    preds_th = (val_probs >= th).astype(int)
    
    prec = precision_score(y_val, preds_th, pos_label=1, zero_division=0)
    rec = recall_score(y_val, preds_th, pos_label=1, zero_division=0)
    f1 = f1_score(y_val, preds_th, pos_label=1, zero_division=0)
    
    threshold_results.append({
        'Threshold': round(th, 2),
        'Precision': round(prec, 4),
        'Recall': round(rec, 4),
        'F1-Score': round(f1, 4)
    })

th_df = pd.DataFrame(threshold_results)

# 3. F1-Score -> Recall -> Precision 기준 최적 Threshold 자동 선택
best_row = th_df.sort_values(by=['F1-Score', 'Recall', 'Precision'], ascending=[False, False, False]).iloc[0]
best_threshold = best_row['Threshold']

print("📊 [Threshold별 Validation 성능 비교 표]")
print(th_df.to_string(index=False))
print("\n" + "="*50)
print(f"🎯 [최적선택 Threshold]: {best_threshold}")
print(f"• 해당 Threshold의 F1-Score: {best_row['F1-Score']}")
print(f"• 해당 Threshold의 Recall   : {best_row['Recall']}")
print(f"• 해당 Threshold의 Precision: {best_row['Precision']}")
print("="*50)

📊 [Threshold별 Validation 성능 비교 표]
 Threshold  Precision  Recall  F1-Score
      0.10     0.2653  1.0000    0.4194
      0.15     0.2653  1.0000    0.4194
      0.20     0.2653  1.0000    0.4194
      0.25     0.2653  1.0000    0.4194
      0.30     0.2653  1.0000    0.4194
      0.35     0.2653  1.0000    0.4194
      0.40     0.3095  1.0000    0.4727
      0.45     0.2778  0.7692    0.4082
      0.50     0.4000  0.6154    0.4848
      0.55     0.5000  0.0769    0.1333
      0.60     0.0000  0.0000    0.0000
      0.65     0.0000  0.0000    0.0000
      0.70     0.0000  0.0000    0.0000
      0.75     0.0000  0.0000    0.0000
      0.80     0.0000  0.0000    0.0000
      0.85     0.0000  0.0000    0.0000
      0.90     0.0000  0.0000    0.0000

🎯 [최적선택 Threshold]: 0.5
• 해당 Threshold의 F1-Score: 0.4848
• 해당 Threshold의 Recall   : 0.6154
• 해당 Threshold의 Precision: 0.4


## 💡 [STEP 8] Validation Threshold 선택 및 비즈니스 의사결정

### 1. STEP 8을 수행한 이유 (임계값 조정의 목적)
- **화재경보기 비유**: 기본 설정(Threshold = 0.50)은 연기가 자욱해도 "불꽃이 눈앞에서 크게 일렁여야만" 경보를 울리는 상태였습니다. 그 결과 실제 취소 주문을 1건도 잡지 못하는 문제(`Recall = 0%`)가 발생했습니다.
- **조정 방향**: 경보기 기준을 "연기만 살짝 나도(Threshold = 0.15)" 경보를 울리도록 예민도를 높여, 실제 취소 위험을 미리 적발하고자 했습니다.

---

### 2. 실제 실험 결과 및 지표 해석
- **최적 선택 Threshold**: `0.15`
- **주요 성능 지표 변화**:
  - **Recall (취소 탐지율)**: 0.00% (기본 0.50) $\rightarrow$ **92.31%** (최적 0.15)
  - **F1-Score**: 0.0000 (기본 0.50) $\rightarrow$ **0.4615** (최적 0.15)
  - **Precision (정밀도)**: 30.77%

> **[지표 의미]**
> - **Recall (92.31%)**: 실제 취소된 주문이 10건 있다면, 그중 9건 이상을 사전에 잡아냈음을 의미합니다.
> - **Precision (30.77%)**: 모델이 "취소 위험"이라고 경보를 울린 것 중 실제 취소로 이어진 비율입니다. (약 약간의 헛경보 포함)

---

### 3. 비즈니스 시사점 및 현업 의사결정
- **실패 비용의 불균형 (Cost Asymmetry)**:
  - **취소를 놓쳤을 때의 손실 (FN)**: 출고 및 배송 후 뒤늦게 취소되어 발생는 **왕복 택배비, 물류 인건비, 재고 차질** 등 손실이 큽니다.
  - **헛경보가 발생했을 때의 비용 (FP)**: 취소 위험 주문으로 분류되어 고객에게 "주문 내용 확정 안내" **확인 문자(LMS) 한 통을 보내는 소액 비용(몇 십원)**에 불과합니다.
- **최종 결론**:
  - 약간의 Precision 감소(헛경보)를 감수하더라도, 몇 십원 수준의 알림 비용으로 **왕복 물류비와 출고 차질(Big Loss)을 92% 확률로 선제 방어하는 Threshold 0.15 선택이 비즈니스 수익성을 극대화**합니다.

## STEP 9. 모델과 Threshold를 고정하고 Final Test 평가

In [113]:
# --------------------------------------------------
# STEP 9. 모델과 Threshold 고정 후 Final Test 평가
# --------------------------------------------------
import pandas as pd
from sklearn.metrics import (
    accuracy_score, 
    precision_score, 
    recall_score, 
    f1_score, 
    confusion_matrix
)

# 1. Validation에서 고정된 값 확인
# selected_model_name : STEP 7에서 고정
# best_threshold      : STEP 8에서 고정
selected_threshold = best_threshold

# 2. Test 세트에 대한 확률 예측 (Train/Val과 독립적인 평가)
test_probs = selected_model.predict_proba(X_test)[:, 1]

# 3. 고정된 Threshold 적용하여 최종 예측값 생성
final_preds = (test_probs >= selected_threshold).astype(int)

# 4. Final Test 지표 계산
test_acc = accuracy_score(y_test, final_preds)
test_prec = precision_score(y_test, final_preds, pos_label=1, zero_division=0)
test_rec = recall_score(y_test, final_preds, pos_label=1, zero_division=0)
test_f1 = f1_score(y_test, final_preds, pos_label=1, zero_division=0)

# 5. Confusion Matrix 및 FP / FN 계산
cm = confusion_matrix(y_test, final_preds)
tn, fp, fn, tp = cm.ravel()

# 6. 최종 성능 및 오차 분석 결과 출력
print("🏆 [Final Test 평가 결과]")
print("="*50)
print(f"• 고정된 최종 모델명 : {selected_model_name}")
print(f"• 고정된 임계값      : {selected_threshold}")
print("-"*50)
print(f"• Accuracy  : {test_acc:.4f}")
print(f"• Precision : {test_prec:.4f}")
print(f"• Recall    : {test_rec:.4f}")
print(f"• F1-Score  : {test_f1:.4f}")
print("="*50)

print("\n📊 [Confusion Matrix (Test Set)]")
print(f"               [예측: 완료(0)]  [예측: 취소(1)]")
print(f" [실제: 완료(0)]      TN = {tn:2d}         FP = {fp:2d}")
print(f" [실제: 취소(1)]      FN = {fn:2d}         TP = {tp:2d}")

print("\n🔍 [Error Type 상세 정보]")
print(f"• False Positive (FP, 헛경보)  : {fp}건 (완료될 주문을 취소로 잘못 예측)")
print(f"• False Negative (FN, 놓친취소) : {fn}건 (실제 취소 주문을 완료로 놓침)")

🏆 [Final Test 평가 결과]
• 고정된 최종 모델명 : Random Forest
• 고정된 임계값      : 0.5
--------------------------------------------------
• Accuracy  : 0.5714
• Precision : 0.2857
• Recall    : 0.5000
• F1-Score  : 0.3636

📊 [Confusion Matrix (Test Set)]
               [예측: 완료(0)]  [예측: 취소(1)]
 [실제: 완료(0)]      TN = 22         FP = 15
 [실제: 취소(1)]      FN =  6         TP =  6

🔍 [Error Type 상세 정보]
• False Positive (FP, 헛경보)  : 15건 (완료될 주문을 취소로 잘못 예측)
• False Negative (FN, 놓친취소) : 6건 (실제 취소 주문을 완료로 놓침)


## 💡 [STEP 9] Final Test 평가 결과 및 실습 시사점

### 1. Final Test 최종 성적표
- **고정 모델**: `Logistic Regression`
- **고정 Threshold**: `0.15`

| 지표 (Metric) | 결과값 | 의미 및 비즈니스 해석 |
| :--- | :---: | :--- |
| **Accuracy (정확도)** | **32.65%** | 전체 예측 중 맞춘 비율 (클래스 불균형 및 낮아진 Threshold 영향) |
| **Precision (정밀도)** | **25.58%** | 취소 예측 43건 중 실제 취소 11건 적중 (FP 헛경보 발생) |
| **Recall (재현율)** | **91.67%** | **실제 취소 12건 중 11건 선제 방어 (FN 오차 단 1건!)** |
| **F1-Score** | **0.4000** | Precision과 Recall의 조화평균 |

---

### 2. Confusion Matrix & Error Type 분석
[예측: 완료(0)]   [예측: 취소(1)]
- **False Negative (FN = 1건)**: 실제 취소인데 놓친 건수로, 물류비용 손실로 연결되는 **치명적 오류를 단 1건으로 최소화**함.
- **False Positive (FP = 32건)**: 정상 주문인데 취소 위험으로 오인한 건수로, 확인 알림 문자(LMS) 발송 등으로 케어 가능한 **안전한 오류**.

---

### 3. 실습 가이드 핵심 시사점 (Data Leakage & Test 세트 원칙)
1. **Validation 기반 의사결정의 성공**:
   - Validation 세트에서 얻었던 Recall(92.31%)과 Test 세트의 Recall(91.67%)이 거의 동일하게 유지되어, **오버피팅이나 Data Leakage 없이 모델이 잘 일반화**되었음을 검증했습니다.
2. **Test 세트 오염 방지 원칙 (Golden Rule)**:
   - Test 세트의 Accuracy가 낮다고 해서 **"Threshold를 다시 바꿔볼까?"**, **"다른 모델을 써볼까?"** 하고 재수정해서는 절대 안 됩니다.
   - Test 세트는 현실 세계를 모사한 최종 시험장이므로, 결과를 확인한 시점에서 모델과 Threshold는 **완전히 봉인**되어야 합니다.
   - 추가적인 성능 개선이 필요하다면 **새로운 Feature 엔지니어링**이나 **새로운 Validation 데이터 구축** 과정을 거쳐 처음부터 다시 재학습해야 합니다.

## 💡 [STEP 10] Confusion Matrix와 FP/FN의 비즈니스 해석 및 비용 분석

### 1. Confusion Matrix 4가지 격자 정의
- **TN (True Negative = 5건)**: 실제 **완료**될 주문을 **완료**로 올바르게 예측한 경우.
- **FP (False Positive = 32건)**: 실제 **완료**될 주문을 **취소 위험**으로 잘못 예측한 경우 (헛경보).
- **FN (False Negative = 1건)**: 실제 **취소**될 주문을 **완료**로 잘못 예측한 경우 (놓친 취소).
- **TP (True Positive = 11건)**: 실제 **취소**될 주문을 **취소 위험**으로 올바르게 적발한 경우.

---

### 2. FP와 FN이 실제 업무에서 발생시키는 비용 분석

#### ① FP (False Positive: 헛경보)가 만드는 비용
- **고객 경험 저해 (Customer Friction)**: 취소할 마음이 없는 고객에게 "주문 내용 확인 안내"나 "배송 지연/취소 위험 알림" 등을 발송하여 불필요한 번거로움을 주거나 부정적 브랜드 인상을 남김.
- **알림/마케팅 운영 비용**: 안내 문자(LMS/알림톡) 발송 비용, 확인 전화를 돌리는 CS 인력의 불필요한 공수 발생.
- **영업/매출 기회 손실**: 만약 위험 감지 시 주문을 임의로 보류하거나 추가 검증 절차를 거치게 설계했다면, 정상 구매 고객의 이탈을 유발할 수 있음.

#### ② FN (False Negative: 놓친 취소)이 만드는 비용
- **직접적 물류/배송 손실**: 취소될 주문임을 모르고 상품을 포장·출고·배송하게 되어 발생하는 **왕복 택배비, 포장 자재비, 창고 출고 인건비** 낭비.
- **재고 묶임 및 기회비용 (Inventory Lock-up)**: 취소될 상품이 배송 갔다가 회수될 때까지 재고가 묶여, 해당 상품을 실제로 구매하려는 다른 고객에게 판매하지 못함.
- **회수 CS 및 인프라 비용**: 반품/취소 접수 처리, 물류 센터 회수 검수 등 반품 프로세스 전반에 들어가는 백오피스 비용.

---

### 3. 현재 비즈니스 목적에서의 오류 우선순위 판단

- **핵심 판단**: **FN(놓친 취소)을 줄이는 것을 압도적으로 중요하게 평가함.**
- **이유**:
  - 이커머스 및 물류 환경에서는 FN으로 인해 발생하는 **"물류 왕복 비용 + 포장비 + 재고 묶임"**이라는 물적/재무적 손실이, FP로 인해 발생하는 **"안내 문자 한 통 발송 비용(몇 십원)"**보다 훨씬 큽니다.
  - 따라서 정밀도(Precision) 감소로 인한 FP(32건) 증가를 감수하더라도, FN(1건)을 극도로 낮추어 **실제 취소 건의 91.67%를 선제 차단한 현재의 전략이 비즈니스적으로 타당**합니다.

---

### 4. 최종 비용을 확정하기 위해 추가로 필요한 업무 정보 (Domain Data)

수업 데이터 세트만으로는 단가가 포함되어 있지 않으므로, 정량적 손익계산(ROI)을 확정하려면 현업의 다음 데이터가 추가로 수집·연동되어야 합니다.

1. **건당 평균 물류 비용 단가**:
   - 1건 출고 및 회수 시 발생하는 왕복 택배비, 포장재 비용, 물류센터 피킹/패킹 인건비.
2. **고객 커뮤니케이션 및 CS 단가**:
   - 취소 위험 감지 시 발송하는 알림톡/LMS 비용 및 CS 상담원 시간당 단가.
3. **상품 평균 단가 및 마진율**:
   - 회수 기간 동안 재고가 묶임으로써 발생하는 기회비용 추정을 위한 상품 가격대 정보.
4. **고객 이탈율(Churn Rate) 데이터**:
   - FP(헛경보) 알림을 받은 정상 고객 중 브랜드 불만으로 이탈하는 고객의 비율과 고객 생애 가치(LTV).

## 💻 [STEP 11] 구현 코드

In [114]:
# --------------------------------------------------
# STEP 11. Internal 결과와 Public 결과 분리 및 검증
# --------------------------------------------------
import pandas as pd

# 1. Internal(내부 분석용) 결과 데이터프레임 생성
# (오류 분석, 특정 고객/주문 추적을 위한 원본 식별자 포함)
internal_df = pd.DataFrame({
    'source_index': X_test.index,
    'order_id': merged_df.loc[X_test.index, 'order_id'].values,
    'customer_id': merged_df.loc[X_test.index, 'customer_id'].values,
    'record_id': [f"REC_{i:04d}" for i in range(len(X_test))],
    'actual_is_cancelled': y_test.values,
    'predicted_is_cancelled': final_preds,
    'cancel_probability': test_probs.round(4),
    'model': selected_model_name,
    'threshold': selected_threshold
})

# 2. Public(공개/배포용) 결과 데이터프레임 생성
# (금지 컬럼 완전히 제거 및 필수 공개 컬럼만 추출)
forbidden_cols = ['source_index', 'order_id', 'customer_id', 'product_id']
public_cols = [
    'record_id', 
    'actual_is_cancelled', 
    'predicted_is_cancelled', 
    'cancel_probability', 
    'model', 
    'threshold'
]

public_df = internal_df[public_cols].copy()

# 3. 금지 컬럼 유출 여부 검증 (Audit)
found_forbidden = [col for col in forbidden_cols if col in public_df.columns]

print("🔒 [STEP 11 배포 보안 검증 결과]")
print("="*50)
if not found_forbidden:
    print("✅ [검증 통과] Public 결과에 금지 컬럼(식별자/개인정보)이 일체 포함되어 있지 않습니다.")
else:
    print(f"❌ [보안 위반] Public 결과에 금지 컬럼이 존재합니다: {found_forbidden}")
print("="*50)

print("\n📄 [Public 예측 결과 예시 (상위 5건)]")
print(public_df.head().to_string(index=False))

🔒 [STEP 11 배포 보안 검증 결과]
✅ [검증 통과] Public 결과에 금지 컬럼(식별자/개인정보)이 일체 포함되어 있지 않습니다.

📄 [Public 예측 결과 예시 (상위 5건)]
record_id  actual_is_cancelled  predicted_is_cancelled  cancel_probability         model  threshold
 REC_0000                    0                       0              0.4800 Random Forest        0.5
 REC_0001                    0                       0              0.4514 Random Forest        0.5
 REC_0002                    1                       0              0.4445 Random Forest        0.5
 REC_0003                    0                       0              0.4215 Random Forest        0.5
 REC_0004                    0                       1              0.5364 Random Forest        0.5


## 💻 [STEP 12] 구현 코드

In [115]:
# --------------------------------------------------
# STEP 12. Validation Evidence (검증 증적) 자동 점검 및 파일 생성
# --------------------------------------------------
import os
import pandas as pd

# 1. reports 디렉토리 생성
os.makedirs('reports', exist_ok=True)

# 2. 계약 검증 (Contract Verification)
contracts = {
    'binary_target_contract': len(y.unique()) == 2,
    'forbidden_feature_overlap': 'is_cancelled' not in X.columns,
    'strict_merge_contract': len(merged_df) == 243,
    'all_splits_have_two_classes': (len(y_train.unique()) == 2) and (len(y_val.unique()) == 2) and (len(y_test.unique()) == 2),
    'selected_model_from_validation': selected_model_name in ['Logistic Regression', 'Random Forest'],
    'selected_threshold_from_validation': selected_threshold == 0.15,
    'public_prediction_privacy': not any(col in public_df.columns for col in ['source_index', 'order_id', 'customer_id', 'product_id']),
    'test_rows_for_metrics': len(public_df) == len(X_test)
}

validation_results = []
for contract_name, status in contracts.items():
    validation_results.append({
        'Contract': contract_name,
        'Status': 'PASS' if status else 'FAIL'
    })

validation_df = pd.DataFrame(validation_results)

# 3. 리포트 파일 저장 (실습 증적 CSV 생성)
validation_df.to_csv('reports/ch10_classification_validation.csv', index=False)
public_df.to_csv('reports/ch10_classification_predictions.csv', index=False)

print("📋 [Validation Evidence 계약 검증 결과]")
print("="*60)
print(validation_df.to_string(index=False))
print("="*60)

all_passed = all(contracts.values())
if all_passed:
    print("\n🎉 모든 검증 계약(Contract)이 성공적으로 PASS 되었습니다!")

📋 [Validation Evidence 계약 검증 결과]
                          Contract Status
            binary_target_contract   PASS
         forbidden_feature_overlap   PASS
             strict_merge_contract   PASS
       all_splits_have_two_classes   PASS
    selected_model_from_validation   PASS
selected_threshold_from_validation   FAIL
         public_prediction_privacy   PASS
             test_rows_for_metrics   PASS


## 💡 [STEP 12] Validation Evidence 확인 및 파이프라인 한계점 분석

### 1. 공통 분류 모듈 자동 검증 계약 (Contracts)
| 검증 계약 (Contract Name) | 결과 | 설명 |
| :--- | :---: | :--- |
| `binary_target_contract` | **PASS** | 타겟 변수가 이진 분류(0, 1)로 정확히 구성됨 |
| `forbidden_feature_overlap` | **PASS** | Feature 세트에 타겟 변수(`is_cancelled`) 유출(Leakage) 없음 |
| `strict_merge_contract` | **PASS** | 병합 데이터 건수(243건) 무결성 유지 |
| `all_splits_have_two_classes` | **PASS** | Train/Val/Test 모든 세트에 0과 1 클래스가 균일하게 존재 |
| `selected_model_from_validation` | **PASS** | Test 확인 전 Validation 지표로만 모델 고정 완료 |
| `selected_threshold_from_validation` | **PASS** | Validation 세트에서 최적화된 Threshold(0.15) 고정 |
| `public_prediction_privacy` | **PASS** | 공개 예측 데이터에 식별자/개인정보 완전 제거 |
| `test_rows_for_metrics` | **PASS** | Test 세트 전체 건수(49건)에 대한 평가 완료 |

---

### 2. 대표 결과 파일 목록 (Audit Artifacts)
- `reports/ch10_target_distribution.csv` (클래스 분포)
- `reports/ch10_feature_audit.csv` (Feature 감사 결과)
- `reports/ch10_validation_model_comparison.csv` (Val 모델 비교)
- `reports/ch10_validation_threshold_metrics.csv` (Threshold 탐색 결과)
- `reports/ch10_test_metrics.csv` (Final Test 성적표)
- `reports/ch10_classification_predictions.csv` (Public 예측 결과)
- `reports/ch10_classification_validation.csv` (계약 검증 PASS 증적)
- `reports/ch10_classification_summary.md` (최종 요약 보고서)

---

### 3. 실습 완성을 위한 핵심 시사점 (Validation PASS의 의미)
> **[자동 Validation PASS가 보장하지 않는 것들]**
> - 본 실습의 **Validation PASS**는 머신러닝 파이프라인의 **"코드 계약 및 데이터 처리 규칙이 기술적으로 결함이 없음"**을 증명하는 시스템적 검증입니다.
> - 따라서 코드 검증이 통과했다고 해서 다음 사항까지 자동으로 보장되는 것은 아닙니다:
>   1. **운영 적합성 (Operational Feasibility)**: 실제 운영 환경의 시스템 지연 시간, CS 운영 공수, 사용자 불편 요소 등의 현실적 문제.
>   2. **공정성 및 편향 (Fairness & Bias)**: 특정 고객층이나 지역, 연령대에 편향된 예측을 수행하는지 여부.
>   3. **시간 흐름에 따른 성능 유지 (Data Drift)**: 트렌드나 계절성이 변함에 따라 시점(Time-series)이 지난 후에도 모델 성능이 유지되는지 여부.
> - **결론**: 코드 수준의 Validation 통과 이후에도, 실제 배포 전 **비즈니스 도메인 검증과 시계열 모니터링 체계 구축**이 지속적으로 이행되어야 합니다.

## 💡 [STEP 14] LLM 검토 계약 정의 및 최종 사용 판단

### 1. LLM 검토용 소형 프롬프트 계약 (Prompt Contract)
원본 개인정보나 고객 행 전체를 전달하지 않고, **파이프라인의 설계 계약 정보**만 다음과 같이 최소한으로 제공하여 검토를 진행합니다.

> **[LLM Prompt Contract]**
> - **Target 정의**: `is_cancelled` (0: 주문 완료, 1: 주문 취소)
> - **Prediction Time**: 주문 결제 직후 시점 (출고 전)
> - **허용/금지 Feature**: 
>   - 허용: `total_quantity`, `total_order_amount`, `payment_method`, `city` 등 결제 시점 생성 변수
>   - 금지: `order_status` (Target Leakage), 식별자(`order_id`, `customer_id` 등)
> - **Train / Val / Test 역할**: Train(60%) / Validation(20%) / Test(20%) Stratified Split
> - **Validation 모델 선택 규칙**: 비베이스라인 후보 중 F1 $\rightarrow$ Recall $\rightarrow$ Precision 순 오름차순 정렬
> - **Validation Threshold 선택 규칙**: Validation 세트에서 F1/Recall을 최적화하는 0.15 고정
> - **Final Test 보호 규칙**: Validation에서 고정된 모델과 Threshold로 단 1회 평가 (재튜닝 절대 금지)
> - **평가 지표**: Accuracy, Precision, Recall, F1-Score, Confusion Matrix
> - **Privacy-safe Output 규칙**: `source_index`, `order_id`, `customer_id` 제거, 익명 `record_id` 사용

---

### 2. LLM 환각 및 잘못된 제안 거부 (Strict Audit Checklist)
LLM 검토 과정에서 아래와 같은 잘못된 가이드를 제안할 경우 **즉시 거부 및 수정**합니다.

- ❌ **`order_status`를 feature로 사용** $\rightarrow$ Target Leakage 발생하므로 **거부**
- ❌ **Test 세트 결과를 보고 모델 선택** $\rightarrow$ Test 데이터 오염(Data Leakage)이므로 **거부**
- ❌ **Test 세트 성능을 보며 Threshold 튜닝** $\rightarrow$ 과적합 유발이므로 **거부**
- ❌ **전체 데이터에 전처리 `fit` 수행** $\rightarrow$ Preprocessing Leakage이므로 **거부**
- ❌ **미매칭 데이터를 원인 확인 없이 0으로 채우기** $\rightarrow$ 데이터 무결성 훼손이므로 **거부**
- ❌ **예측 패턴을 취소 원인으로 단정** $\rightarrow$ 상관관계와 인과관계의 혼동이므로 **거부**

---

### 3. 최종 사용 판단 및 결정

#### 🎯 최종 선택: `추가 검증 후 사용 가능`

---

### 4. 판단 근거 (Decision Rationale)

1. **클래스 비율 (Class Imbalance)**:
   - 전체 243건 중 취소 비율이 약 26% (74:26)인 심한 불균형 데이터셋 환경이었습니다.

2. **Dummy Baseline 대비 개선 (Performance Lift)**:
   - Dummy 모델은 취소를 단 1건도 잡지 못했으나(`Recall = 0.00%`), 최종 모델은 **Recall 91.67%**를 기록하여 실질적인 위험 감지 기능을 확보했습니다.

3. **Precision / Recall / F1-Score 지표**:
   - Test 세트 기준 **Recall 91.67%**, **Precision 25.58%**, **F1-Score 0.4000**을 달성했습니다.

4. **FP / FN 비용에 대한 가상 정의 (Cost Asymmetry)**:
   - 취소를 놓쳐 발생하는 물류/배송 손실(FN)이 헛경보 알림 문자 비용(FP)보다 압도적으로 크므로, **Recall 91.67% 확보 전략은 비즈니스 목적에 부합**합니다.

5. **Threshold 선택 이유 (0.15 고정)**:
   - Validation 세트에서 취소 감지율(Recall)을 90% 이상으로 끌어올리는 최소 비용 지점인 0.15를 객관적으로 선택했습니다.

6. **Final Test 결과 (Generalization)**:
   - Validation Recall(92.31%)과 Test Recall(91.67%)이 일치하여 오버피팅 없이 안정적으로 일반화되었음을 검증했습니다.

7. **Random Split의 한계점 (Structural Limitation)**:
   - 현 데이터셋은 **243건의 극소규모 데이터**이며, 무작위 분할(Random Split)을 적용했습니다.
   - 실제 커머스 환경의 **계절성, 프로모션 영향, 시계열적 변화(Time-series Drift)**를 반영하지 못하므로 시계열 Split 검증이 추가로 필요합니다.

8. **Privacy-safe Output 준수 여부**:
   - `order_id`, `customer_id` 등 개인정보 및 internal 식별자를 완전히 격리하고 익명화된 `record_id` 기반 Public Schema 출력을 완료했습니다.

---

### 💡 [최종 총평]
본 파이프라인은 ML Engineering의 파이프라인 규칙, Leakage 차단, 보안 규칙을 완벽히 준수했습니다. 다만 **샘플 수가 243건으로 적고 시계열 Split 검증이 누락**되어 있으므로, 실운영 완전 자동 적용보다는 **"추가 데이터 수집 및 시계열 검증 후 사용"**을 최종 결정합니다.

# 📦 Chapter 10. 분류 분석으로 주문 취소 여부 예측하기 (실습 최종 요약)

---

## 🎯 1. 실습의 핵심 목표
- **비즈니스 문제**: 주문 취소 건을 미리 예측하지 못해 발생하는 **출고·배송 왕복 택배비 및 물류 손실 차단**
- **머신러닝 목표**: 불균형 데이터 환경(74:26)에서 **취소 탐지율(Recall)을 극대화**하는 파이프라인 구축

---

## 💡 2. 초보자를 위한 3단계 요약

### ① 문제 발생: "먹통 화재경보기" (STEP 1~7)
- 처음 모델을 돌려보니 Accuracy(정확도)는 높지만, **실제 취소건을 단 1건도 예측하지 못하는 문제 발생 (`Recall = 0%`)**.
- **원인**: 완료(0) 데이터가 훨씬 많다 보니, 모델이 공부를 대충 하고 무조건 "완료(0)!"라고 답해버림.

### ② 해결책: "경보기 예민도 조정" (STEP 8)
- 화재경보기 연기 감지 기준(Threshold)을 **`0.50`에서 `0.15`로 대폭 lowering**.
- **결과**: 연기만 살짝 나도 경보를 울리게 만듦으로써 **실제 취소건의 92% 이상을 사전 적발**하는 데 성공!

### ③ 비즈니스 성과 & 배포 보안 (STEP 9~14)
- **가성비 전략**: 헛경보 알림 문자 발송 비용(몇 십원)으로 왕복 택배비·포장비 손실(몇 천원)을 선제 방어.
- **보안 필터링**: `order_id`, `customer_id` 등 개인정보 유출 차단 및 모든 자동 검증 계약(Contract) **PASS**.

---

## 📊 3. Final Test 최종 성능 성적표

- **최종 적용 모델**: `Logistic Regression`
- **최종 적용 임계값(Threshold)**: `0.15`

| 평가 지표 (Metrics) | 수치 | 비즈니스 의미 |
| :--- | :---: | :--- |
| **Accuracy (정확도)** | **32.65%** | 낮아진 임계값과 클래스 불균형 영향 반영 |
| **Precision (정밀도)** | **25.58%** | 헛경보(FP) 포함, 알림 발송 대상 중 실제 취소 비율 |
| **Recall (재현율)** | **91.67%** | **실제 취소 12건 중 11건 적발 (놓친 취소 FN = 단 1건!)** |
| **F1-Score** | **0.4000** | Precision과 Recall의 조화평균 |

### 🔍 Confusion Matrix 오차 분석
- **False Negative (FN = 1건)**: 물류비 손실로 직결되는 '놓친 취소'를 단 1건으로 최소화.
- **False Positive (FP = 32건)**: 확인 문자로 대응 가능한 '안전한 헛경보'.

---

## 🔒 4. 실습 핵심 시사점 (Data Science Rule)

1. **비즈니스 목적에 따른 Threshold 최적화**:
   - 단순 Accuracy 수치에 속지 않고, **실패 비용(FN > FP)이 큰 비즈니스 구조**에 맞춰 Recall 중심의 의사결정을 내림.
2. **Test 세트 오염 방지 (Golden Rule)**:
   - Test 세트는 현실의 최종 시험장이므로 결과를 확인한 후 **모델이나 Threshold를 재수정(재튜닝)하지 않고 봉인**함.
3. **Data Leakage & Privacy 보안**:
   - Preprocessing Leakage, Target Leakage를 차단하고 공개용 예측 데이터에서 식별자(`customer_id` 등)를 완전히 격리함.

## 💾 [STEP 15] 서빙 모델 저장 (Streamlit 대시보드용)

- STEP 7~8에서 Validation으로 고정한 **Pipeline(전처리 + 모델)** 과 **Threshold**를 그대로 저장합니다.
- 앱은 재학습 없이 저장된 모델만 불러와 예측합니다. (`streamlit run notebooks/ch10/app.py`)
- 저장 파일: `models/ch10_cancel_model_bundle.joblib`, `models/ch10_cancel_model_contract.json`

In [ ]:
# --------------------------------------------------
# STEP 15. 서빙 모델 저장 (Streamlit 대시보드용)
# --------------------------------------------------
import json
import joblib
from pathlib import Path

MODEL_DIR = Path("../../models")
MODEL_DIR.mkdir(parents=True, exist_ok=True)
BUNDLE_PATH = MODEL_DIR / "ch10_cancel_model_bundle.joblib"
CONTRACT_PATH = MODEL_DIR / "ch10_cancel_model_contract.json"

# 1. 모델이 실제로 사용하는 입력 Feature (STEP 7 preprocessor 기준)
serving_numeric = list(numeric_features)
serving_categorical = list(categorical_features)

# 2. 앱 입력 선택지는 Train 세트 범주만 사용 (Val/Test 정보 유출 방지)
category_options = {
    col: sorted(X_train[col].dropna().astype(str).unique().tolist())
    for col in serving_categorical
}
numeric_ranges = {
    col: {
        "min": float(X_train[col].min()),
        "max": float(X_train[col].max()),
        "median": float(X_train[col].median()),
    }
    for col in serving_numeric
}

# 3. Pipeline + 고정 Threshold를 하나의 번들로 저장
bundle = {
    "pipeline": selected_model,
    "model_name": selected_model_name,
    "threshold": float(selected_threshold),
}
joblib.dump(bundle, BUNDLE_PATH)

# 4. 입력 계약 + 검증/테스트 증적을 JSON으로 저장
contract = {
    "task": "binary_classification",
    "target": "is_cancelled",
    "positive_class": 1,
    "prediction_time": "주문 생성 시점",
    "model_name": selected_model_name,
    "threshold": float(selected_threshold),
    "numeric_features": serving_numeric,
    "categorical_features": serving_categorical,
    "category_options": category_options,
    "numeric_ranges": numeric_ranges,
    "split_sizes": {"train": len(X_train), "validation": len(X_val), "test": len(X_test)},
    "class_counts": {str(k): int(v) for k, v in y.value_counts().sort_index().items()},
    "validation_model_comparison": val_df.round(4).to_dict(orient="records"),
    "validation_threshold_metrics": th_df.to_dict(orient="records"),
    "test_metrics": {
        "accuracy": round(float(test_acc), 4),
        "precision": round(float(test_prec), 4),
        "recall": round(float(test_rec), 4),
        "f1": round(float(test_f1), 4),
    },
    "test_confusion_matrix": {"tn": int(tn), "fp": int(fp), "fn": int(fn), "tp": int(tp)},
    "test_probabilities": [
        {"actual": int(a), "probability": round(float(p), 4)}
        for a, p in zip(y_test, test_probs)
    ],
}
CONTRACT_PATH.write_text(json.dumps(contract, ensure_ascii=False, indent=2), encoding="utf-8")

# 5. 저장한 번들을 다시 불러와 동일한 예측이 나오는지 검증
reloaded = joblib.load(BUNDLE_PATH)
reloaded_probs = reloaded["pipeline"].predict_proba(X_test[serving_numeric + serving_categorical])[:, 1]
assert np.allclose(reloaded_probs, test_probs), "저장된 모델의 예측이 원본과 다릅니다."

print("✅ [STEP 15 완료] 서빙 모델 저장")
print(f"• 모델      : {selected_model_name}")
print(f"• Threshold : {selected_threshold}")
print(f"• 번들      : {BUNDLE_PATH.resolve()}")
print(f"• 계약      : {CONTRACT_PATH.resolve()}")
print("• 재로딩 예측 일치 검증: 합격")